linera-explorer/src/entrypoint.rs

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use super::js_utils::{getf, js_to_json, setf, SER};
+
+use serde::Serialize;
+use serde_json::Value;
+use serde_wasm_bindgen::from_value;
+use wasm_bindgen::prelude::*;
+
+/// Recursively forge query argument
+fn forge_arg_type(arg: &Value, non_null: bool) -> Option<String> {
+    match arg["kind"].as_str() {
+        Some("SCALAR") => {
+            if non_null {
+                Some(format!("{}", arg["_input"]))
+            } else {
+                arg.get("_input").map(|input| format!("{}", input))
+            }
+        }
+        Some("NON_NULL") => forge_arg_type(&arg["ofType"], true),
+        Some("LIST") => {
+            let args: Vec<String> = arg["_input"]
+                .as_array()
+                .unwrap()
+                .iter()
+                .filter_map(|x| forge_arg_type(x, false))
+                .collect();
+            Some(format!("[{}]", args.join(", ")))
+        }
+        Some("ENUM") => arg["_input"].as_str().map(|x| x.to_string()),
+        Some("INPUT_OBJECT") => {
+            let args: Vec<String> = arg["inputFields"]
+                .as_array()
+                .unwrap()
+                .iter()
+                .filter_map(|x| {
+                    let name = x["name"].as_str().unwrap();
+                    forge_arg_type(&x["type"], false).map(|arg| format!("{}: {}", name, arg))
+                })
+                .collect();
+            Some(format!("{{{}}}", args.join(", ")))
+        }
+        _ => None,
+    }
+}
+
+fn forge_arg(arg: &Value) -> Option<String> {
+    forge_arg_type(&arg["type"], false).map(|s| {
+        format!(
+            "{}: {}",
+            arg["name"].as_str().expect("name is not a string"),
+            s
+        )
+    })
+}
+
+fn forge_args(args: Vec<Value>) -> String {
+    let args: Vec<String> = args.iter().filter_map(forge_arg).collect();
+    if !args.is_empty() {
+        format!("({})", args.join(","))
+    } else {
+        "".to_string()
+    }
+}
+
+/// Recursively forge query response
+fn forge_response_type(t: &Value, name: Option<&str>, root: bool) -> String {
```

Comment:
We're trying to avoid single-letter variable names in general.

```suggestion
fn forge_response_type(value: &Value, name: Option<&str>, root: bool) -> String {
```